### PR TITLE
Test and optimize Sha256Hash.toString()

### PR DESCRIPTION
**Pull request by schlosna**
## General

`Sha256Hash.toString()` uses expensive `String.format("%02x", byte)` & `StringBuilder` to hex encode the hash digest bytes.

**After this PR**:

==COMMIT_MSG==
Test and optimize Sha256Hash.toString()
==COMMIT_MSG==

**Priority**:

**Concerns / possible downsides (what feedback would you like?)**:

**Is documentation needed?**:

## Compatibility
**Does this PR create any API breaks (e.g. at the Java or HTTP layers) - if so, do we have compatibility?**:

**Does this PR change the persisted format of any data - if so, do we have forward and backward compatibility?**:

**The code in this PR may be part of a blue-green deploy. Can upgrades from previous versions safely coexist? (Consider restarts of blue or green nodes.)**:

**Does this PR rely on statements being true about other products at a deployment - if so, do we have correct product dependencies on these products (or other ways of verifying that these statements are true)?**:

**Does this PR need a schema migration?**

## Testing and Correctness
**What, if any, assumptions are made about the current state of the world? If they change over time, how will we find out?**:

**What was existing testing like? What have you done to improve it?**:

**If this PR contains complex concurrent or asynchronous code, is it correct? The onus is on the PR writer to demonstrate this.**:

**If this PR involves acquiring locks or other shared resources, how do we ensure that these are always released?**:

## Execution
**How would I tell this PR works in production? (Metrics, logs, etc.)**:

**Has the safety of all log arguments been decided correctly?**:

**Will this change significantly affect our spending on metrics or logs?**:

**How would I tell that this PR does not work in production? (monitors, etc.)**:

**If this PR does not work as expected, how do I fix that state? Would rollback be straightforward?**:

**If the above plan is more complex than “recall and rollback”, please tag the support PoC here (if it is the end of the week, tag both the current and next PoC)**:

## Scale
**Would this PR be expected to pose a risk at scale? Think of the shopping product at our largest stack.**:

**Would this PR be expected to perform a large number of database calls, and/or expensive database calls (e.g., row range scans, concurrent CAS)?**:

**Would this PR ever, with time and scale, become the wrong thing to do - and if so, how would we know that we need to do something differently?**:

## Development Process
**Where should we start reviewing?**:

**If this PR is in excess of 500 lines excluding versions lock-files, why does it not make sense to split it?**:

**Please tag any other people who should be aware of this PR**:
@jeremyk-91
@sverma30
@raiju

### CODE REVIEW

File: atlasdb-commons/src/test/java/com/palantir/util/crypto/Sha256HashTest.java [hunk 1]
```diff
@@ -29,9 +34,54 @@ public void testSha256() {
...
+    }
+
+    @ParameterizedTest
+    @ValueSource(strings = {"test", "Hello, World", "12345678901234567890"})
+    public void testComputeHash(String input) {
+        byte[] inputBytes = input.getBytes(StandardCharsets.UTF_8);
+        String expectedSha256 = expectedSha256(inputBytes);
+
+        Sha256Hash hash = Sha256Hash.computeHash(inputBytes);
+        String legacyHexEncoded = legacyHex(
```
Comment (mdaudali) [inline atlasdb-commons/src/test/java/com/palantir/util/crypto/Sha256HashTest.java]: nit: You can use legacyHex(hash.getBytes()) if I've understood the goal, and to _me_, the intent is clearer.

File: atlasdb-commons/src/test/java/com/palantir/util/crypto/Sha256HashTest.java [hunk 2]
```diff
@@ -29,9 +34,54 @@ public void testSha256() {
...
+                Sha256Hash.LOWER_CASE_HEX.decode(hash.serializeToHexString().toLowerCase(Locale.ROOT)));
+        assertThat(hash)
+                .isEqualTo(Sha256Hash.computeHash(inputBytes))
+                .hasSameHashCodeAs(Sha256Hash.computeHash(inputBytes))
+                .isEqualByComparingTo(Sha256Hash.computeHash(inputBytes))
+                .asString()
```
Comment (mdaudali) [inline atlasdb-commons/src/test/java/com/palantir/util/crypto/Sha256HashTest.java:93]: nit: Can the things above the asString and those below be two separate assertions? It's a long enough chain that I missed the asString call and was (briefly) wondered how the types lined up!

Comment (mdaudali) [reply atlasdb-commons/src/test/java/com/palantir/util/crypto/Sha256HashTest.java:93]: Especially as you seem to do (partially) the same assertions with hash.toString anyway, so I'm curious what the difference between the two sets is that you're aiming to defend against.

**Review by mdaudali — APPROVED** [review]
lgtm, 2 non-blocking nits

File: atlasdb-commons/src/test/java/com/palantir/util/crypto/Sha256HashTest.java (hunk 1 above)
Comment (schlosna) [reply atlasdb-commons/src/test/java/com/palantir/util/crypto/Sha256HashTest.java]: Will update

File: atlasdb-commons/src/test/java/com/palantir/util/crypto/Sha256HashTest.java (hunk 2 above)
Comment (schlosna) [reply atlasdb-commons/src/test/java/com/palantir/util/crypto/Sha256HashTest.java:93]: I will make these assertions clearer.

I have a feeling the assertj excavator will end up rewriting `assertThat(hash.toString())` as `assertThat(hash).asString()` anyway.

File: atlasdb-commons/src/test/java/com/palantir/util/crypto/Sha256HashTest.java (hunk 1 above)
Comment (schlosna) [reply atlasdb-commons/src/test/java/com/palantir/util/crypto/Sha256HashTest.java]: updated

File: atlasdb-commons/src/test/java/com/palantir/util/crypto/Sha256HashTest.java (hunk 2 above)
Comment (schlosna) [reply atlasdb-commons/src/test/java/com/palantir/util/crypto/Sha256HashTest.java:93]: split the assertions up and also expanded the inputs

**Pull request merged**